src/Gadfly.jl

```diff
@@ -1102,6 +1114,17 @@ include("geometry.jl")
 include("guide.jl")
 include("statistics.jl")
 
+function apply_theme_transformations!(geom::GeometryElement, stat::StatisticElement, aes::Aesthetics, theme::Theme)
+#     @info "apply_theme_transformations!" geom stat aes theme
+end
+
+function apply_theme_transformations!(geom::Geom.RectangularGeometry, stat::Stat.RectbinStatistic, aes::Aesthetics, theme::Theme)
+
+    @info "apply_theme_transformations! specific to geom::Geom.RectangularGeometry, stat::Stat.RectbinStatistic combination."
+
+	 aes.xmax = aes.xmax .* cx .- theme.bar_spacing
+	 aes.ymax = aes.ymax .* cy .- theme.bar_spacing
+end
```

Comment:
@Mattriks @tlnagy thoughts on this proposed new `apply_theme_transformations!` addition to the existing framework?

seems okay to me, but it's not clear what advantages it confers to just handling this case in the corresponding `render` function.

are there other cases which could be refactored like this?

certainly the `@info` and `@warn` need to go.

Reply:
I naively wondered if it were possible to handle default colour and alpha with this sort of pattern.

Reply:
I'm not liking `apply_theme_transformations!`, starting with the strange looping in `render_prepared()`.    Although `render_prepared` has an argument `layer_stats`, that argument is not used in `render_prepared`, and there is no reason for it to be: the applying of `Stats` is done by `render_prepare()`.

This sort of "code to handle common cases" should be done once there is a good idea of what those common cases are.  For example, over time the `default_aes` list (which appears at the start of render functions for polygon, ribbon, point etc) has become more similar, and this is where "unifying code" might be useful.

Reply:
Regarding the strange looping in `render_prepared()`, I picked a spot between application of statistics and rendering where all of the variables required to demonstrate the concept were readily available; it was by no means a suggestion of final implementation. Commit https://github.com/GiovineItalia/Gadfly.jl/commit/bac3a29bd13ca6cacc45e6fcd22a690a3026fd31 places the `apply_theme_transformations!` methods where I'd expect to find them. However, the commit does not develop the hook call.

Reply:
In developing a new `apply_` function, I would like to see a design outline like this:

```julia
# Minimal Gadfly

# function render_prepare
datas = Gadfly.Data(x=1:10, y=rand(10), xend=1:10, yend=zeros(10))
scales=Dict{Symbol,Gadfly.ScaleElement}(:x=>Scale.x_continuous(), :y=>Scale.y_continuous())
aes = Scale.apply_scales(collect(values(scales)), datas)
coord = Coord.cartesian()
stats = Gadfly.StatisticElement[Stat.xticks(), Stat.yticks()]
Stat.apply_statistics(stats, scales, coord, aes[1])
# stats have been implemented, and the stats vector is no longer needed

# function render_prepared
plot_context = Coord.apply_coordinate(coord, aes, scales)
layer_themes = [Theme(default_color="green"), Theme(default_color="orange")]
geom_ctxs = render.([Geom.point(), Geom.hair()], layer_themes, aes)
guide_ctxs = render.([Guide.xticks(), Guide.yticks()], layer_themes, aes)
compose!(plot_context, (context(), geom_ctxs...))
tbl =  Guide.layout_guides(plot_context, coord, Theme(), [x[1] for x in  guide_ctxs]...)

p = compose!(context(), tbl)
```
with an explanation about why I want to add a new `apply_` function to the above code. Then I would ask about timings. Also, I would prefer to see such a new function in a separate PR.

Reply:
I will withdraw the commits for the `apply_theme_transformations!` concept and place them in [feature/rect-apply_theme_transformations](https://github.com/CiaranOMara/Gadfly.jl/tree/feature/rect-apply_theme_transformations). When I have the opportunity, I'll open another PR with a satisfactory explanation and design.

Reply:
agreed that a major change like that should go in a separate PR.  ashame though to have so much code duplication now between render(RectGeo and render(RectBinGeo.

do the regression tests still pass with things as they stand now?

i'd like to see all four commits squashed into a single one.  no need to clutter the git log like this.

Reply:
I think it's better for those two geometries (RectGeo and RectBinGeo) to evolve separately (let's give them the chance to do that).

Reply:
@bjarthur, as of the rollback to https://github.com/GiovineItalia/Gadfly.jl/commit/83399184eeb53924d86546472be67f4f8a790fee there is a regression in `testscripts/rectbin.jl` where `Geom.rect` and `Geom.rectbin` intertwine. I think `Geom.rect` should be removed from the `rectbin` test. I'll have a look at the documentation and make sure the geometries are separated there too. However, this regression alerts me to an important question: what is an appropriate overlap between a `max` and `next min` that share the same value? In the current implementation, is it the stroke that overlaps and not the fill (which I think is appropriate), or does the fill overlap? I'm not sure of the specifics, but suppose it depends on `Compose.jl` and the various backends. I can foresee us encountering undesired artifacts when applying alpha to `Geom.rect` if the nuance of overlap is not correct.

Happy to squash commits once everything looks good.

@Mattriks, the rollback to undoes the changes to `Geom.rectbin` to allow it to evolve separately.

Reply:
is the regression whether the fill and stroke overlap or not?  is there no way to fix it?

Reply:
@bjarthur, I'm confident the regression is due to the exclusion of `theme.bar_spacing` from the new `Geom.rect` implementation.

Below is the `rect.png` test output from https://github.com/GiovineItalia/Gadfly.jl/pull/1263/commits/beee87012e73244ee7be34d6218ad51c508c466a, which is a little fuzzy about the edges.
![rect](https://user-images.githubusercontent.com/24236593/56256202-f89b3800-610a-11e9-8a21-a78a58d165bc.png)

Below is the `rect.pdf` test output zoomed in on (0.5, 0.5) of the plot on the left side.
<img width="317" alt="image" src="https://user-images.githubusercontent.com/24236593/56256264-47e16880-610b-11e9-89fb-570e1d042957.png">

Given these new images, I do not think there is any overlap of fill in the current implementation. However, stroke is currently `stroke(nothing)`.

In terms of regressions, they're isolated to `Geom.rect`. No other geometries derive from `Geom.rect`, other geometries are deriving from `Geom.rectbin`.

Do you think the exclusion of `theme.bar_spacing` from the new `Geom.rect` makes sense?

Reply:
i don't think we can tolerate any regressions until we decide it's time to release 2.0.  if you want to keep this as is, we'll have to put a milestone on it and save it for later.

having a way to slightly adjust the dimensions of the rectangles seems generally useful to me. Theme.bar_spacing doesn't seem the right way to go about it though since that sounds specific to Geom.bar.  but i'd suggest sticking with it for now, and perhaps filing an issue with a milestone 2.0 tag to remind us to change it later.